src/main.zig

```diff
@@ -100,6 +100,20 @@ pub fn main() !void {
     while (true
     // current_epoch_index < TRAINING_EPOCHS
     ) : (current_epoch_index += 1) {
+        // We assume the data is already shuffled so we skip shuffling on the first
+        // epoch. Using a pre-shuffled dataset also gives us nice reproducible results
+        // during the first epoch when trying to debug things.
+        var shuffled_training_data_points = training_data_points;
+        if (current_epoch_index > 0) {
+            // Shuffle the data after each epoch
+            shuffled_training_data_points = try shuffle(random_instance, training_data_points, .{ .allocator = allocator });
```

Comment:
In order to utilize the optimized shuffle algorithms from [`zshuffle`](https://github.com/hmusgrave/zshuffle), @hmusgrave pointed out that we need to pass in a `allocator` otherwise, it just uses Zigs built-in `random_instance.shuffle(...)` implementation ([you can see that logic branch here](https://github.com/hmusgrave/zshuffle/blob/8b0154bea5121208826f0f0c864029e8b0e81446/src/main.zig#L10-L31))